drake/solvers/MobyLCP.cpp

```diff
@@ -0,0 +1,1268 @@
+// Adapted with permission from code by Evan Drumwright
+// (https://github.com/edrumwri).
+
+#include "MobyLCP.h"
+
+#include <Eigen/LU>
+#include <Eigen/SparseCore>
+#include <Eigen/SparseLU>
+
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#include "Optimization.h"
+
+namespace {
+template <typename Derived>
+void selectSubMat(const Eigen::MatrixBase<Derived>& in,
+                  const std::vector<unsigned>& rows,
+                  const std::vector<unsigned>& cols,
+                  Eigen::MatrixXd* out) {
+  const int num_rows = rows.size();
+  const int num_cols = cols.size();
+  out->resize(num_rows, num_cols);
+
+  for (int i = 0; i < num_rows; i++) {
+    const auto row_in = in.row(rows[i]);
+    auto row_out = out->row(i);
+    for (int j = 0; j < num_cols; j++) {
+      row_out(j) = row_in(cols[j]);
+    }
+  }
+}
+
+// TODO sammy this could also use a more efficient implementation.
+template <typename Derived>
+void selectSubVec(const Eigen::MatrixBase<Derived>& in,
+                  const std::vector<unsigned>& rows,
+                  Eigen::VectorXd* out) {
+  const int num_rows = rows.size();
+  out->resize(num_rows);
+  for (int i = 0; i < num_rows; i++) {
+    (*out)(i) = in(rows[i]);
+  }
+}
+
+template <typename DerivedA, typename DerivedB, typename F>
+void transformVecElements(const Eigen::MatrixBase<DerivedA>& in,
+                          Eigen::MatrixBase<DerivedB>* out,
+                          F func) {
+  assert(in.cols() == 1);
+  assert(out->cols() == 1);
+  assert(in.rows() == out->rows());
+  for (int i = 0; i < in.rows(); i++) {
+    (*out)(i) = func(in(i), (*out)(i));
+  }
+}
+
+template <typename Derived>
+Eigen::SparseVector<double> makeSparseVector(
+    const Eigen::MatrixBase<Derived>& in) {
+  assert(in.cols() == 1);
+  Eigen::SparseVector<double> out(in.rows());
+  for (int i = 0; i < in.rows(); i++) {
+    if (in(i) != 0.0) {
+      out.coeffRef(i) = in(i);
+    }
+  }
+  return out;
+}
+
+template <typename Derived>
+Eigen::Index minCoeffIdx(const Eigen::MatrixBase<Derived>& in) {
+  Eigen::Index idx;
+  in.minCoeff(&idx);
+  return idx;
+}
+
+const double NEAR_ZERO = std::sqrt(std::numeric_limits<double>::epsilon());
```

Comment:
epsilon for double is ~2e-16 so `NEAR_ZERO=sqrt(epsilon)` is much larger -- ~1.5e-8